### PR TITLE
improve FetchSourceContext and add unit tests

### PR DESCRIPTION
**Pull request by weizijun**
improve FetchSourceContext and add unit tests

* add null value check in fromXContent, if no null value check, FetchSourceContextTests fromXContent tests will failed. It is similar to GetMappingsResponse.
* fromXContent improve ParsingException's message.
* add FetchSourceContextTests.

### CODE REVIEW

**Comment by weizijun** [comment]
@nik9000 can you help to review this PR?

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceContextTests.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.elasticsearch.test.AbstractSerializingTestCase;
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.containsString;
+
+public class FetchSourceContextTests extends AbstractSerializingTestCase<FetchSourceContext> {
+    @Override
+    protected FetchSourceContext doParseInstance(XContentParser parser) throws IOException {
+        return FetchSourceContext.fromXContent(parser);
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceContextTests.java:25]: When I wrote some tests that recently had the NPE on this sort of thing I would do:
```
        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
       FetchSourceContext result = FetchSourceContext.fromXContent(parser);
       ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.nextToken(), parser);
       return parser;
```

That way I don't have to modify the production code for the test. I think it's *probably* the right thing to do here too.

Comment (weizijun) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceContextTests.java:25]: if fetchSource equals false, the XContent is not object, nor json, it is a value of false.

Comment (nik9000) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FetchSourceContextTests.java:25]: Indeed it does!

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by nik9000** [comment]
@elasticmachine, update branch

**Comment by nik9000** [comment]
@elasticmachine pdate branch

**Comment by nik9000** [comment]
@elasticmachine update branch

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by nik9000** [comment]
>   @elasticsearchmachine elasticsearchmachine merged commit f57a1e7 into elastic:master 19 minutes ago 

Good bot.

**Comment by nik9000** [comment]
Thanks @weizijun! I'll backport this to 7.x now.

**Comment by weizijun** [comment]
> Thanks @weizijun! I'll backport this to 7.x now.

Thanks @nik9000 !

**Pull request merged**